presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControl.java

```diff
@@ -121,14 +120,14 @@ public void checkCanSetUser(Principal principal, String userName)
 
         String principalName = principal.getName();
 
-        for (Pattern pattern : userPatterns) {
-            Matcher matcher = pattern.matcher(principalName);
-            if (matcher.matches()) {
-                for (int i = 1; i <= matcher.groupCount(); i++) {
-                    String extractedUsername = matcher.group(i);
-                    if (userName.equals(extractedUsername)) {
-                        return;
-                    }
+        for (PrincipalUserMatchRule rule : principalUserMatchRules) {
+            Optional<Boolean> allowed = rule.match(principalName, userName);
+            if (allowed.isPresent()) {
+                if (allowed.get()) {
+                    return;
+                }
+                else {
```

Comment:
Nit: the `else` is redundant because the `if` block returns. This can be
```java
if (allowed.get()) {
    return;
}
denySetUser(principal, userName);